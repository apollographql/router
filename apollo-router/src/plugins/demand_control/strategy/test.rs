use apollo_compiler::ExecutableDocument;

use crate::plugins::demand_control::strategy::StrategyImpl;
use crate::plugins::demand_control::test::TestError;
use crate::plugins::demand_control::test::TestStage;
use crate::plugins::demand_control::CostContext;
use crate::plugins::demand_control::DemandControlError;
use crate::services::execution::Request;
use crate::services::subgraph::Response;

/// Test strategy for demand control.
/// Can be configured to fail at different stages of the request processing.
pub(crate) struct Test {
    pub(crate) stage: TestStage,
    pub(crate) error: TestError,
}

impl StrategyImpl for Test {
    fn on_execution_request(&self, request: &Request) -> Result<(), DemandControlError> {
        request.context.extensions().with_lock(|mut lock| {
            let cost_context = lock.get_or_default_mut::<CostContext>();
            match self {
                Test {
                    stage: TestStage::ExecutionRequest,
                    error,
                } => Err(cost_context.result(error.into())),
                _ => Ok(()),
            }
        })
    }

    fn on_subgraph_request(
        &self,
        request: &crate::services::subgraph::Request,
    ) -> Result<(), DemandControlError> {
        request.context.extensions().with_lock(|mut lock| {
            let cost_context = lock.get_or_default_mut::<CostContext>();
            match self {
                Test {
                    stage: TestStage::SubgraphRequest,
                    error,
                } => Err(cost_context.result(error.into())),
                _ => Ok(()),
            }
        })
    }

    fn on_subgraph_response(
        &self,
        _request: &ExecutableDocument,
        response: &Response,
    ) -> Result<(), DemandControlError> {
        response.context.extensions().with_lock(|mut lock| {
            let cost_context = lock.get_or_default_mut::<CostContext>();
            match self {
                Test {
                    stage: TestStage::SubgraphResponse,
                    error,
                } => Err(cost_context.result(error.into())),
                _ => Ok(()),
            }
        })
    }

    fn on_execution_response(
        &self,
        context: &crate::Context,
        _request: &ExecutableDocument,
        _response: &crate::graphql::Response,
    ) -> Result<(), DemandControlError> {
        context.extensions().with_lock(|mut lock| {
            let cost_context = lock.get_or_default_mut::<CostContext>();
            match self {
                Test {
                    stage: TestStage::ExecutionResponse,
                    error,
                } => Err(cost_context.result(error.into())),
                _ => Ok(()),
            }
        })
    }
}
