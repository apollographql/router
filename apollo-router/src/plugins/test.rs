use std::any::TypeId;
use std::ops::Deref;
use std::str::FromStr;
use std::sync::Arc;

use apollo_compiler::validation::Valid;
use serde_json::Value;
use tower::BoxError;
use tower::ServiceBuilder;
use tower_service::Service;

use crate::plugin::DynPlugin;
use crate::plugin::Plugin;
use crate::plugin::PluginInit;
use crate::query_planner::BridgeQueryPlanner;
use crate::services::execution;
use crate::services::http;
use crate::services::router;
use crate::services::subgraph;
use crate::services::supergraph;
use crate::Configuration;
use crate::Notify;

/// Test harness for plugins
/// The difference between this and the regular TestHarness is that this is more suited for unit testing.
/// It doesn't create the entire router stack, and is mostly just a convenient way to call a plugin service given an optional config and a schema.
///
/// Here is a basic example that calls a router service and checks that validates logs are generated for the telemetry plugin.
///
/// ```
///  #[tokio::test(flavor = "multi_thread")]
///     async fn test_router_service() {
///         let test_harness: PluginTestHarness<Telemetry> = PluginTestHarness::builder().build().await;
///
///         async {
///             let mut response = test_harness
///                 .call_router(
///                     router::Request::fake_builder()
///                         .body("query { foo }")
///                         .build()
///                         .expect("expecting valid request"),
///                     |_r| {
///                         tracing::info!("response");
///                         router::Response::fake_builder()
///                             .header("custom-header", "val1")
///                             .data(serde_json::json!({"data": "res"}))
///                             .build()
///                             .expect("expecting valid response")
///                     },
///                 )
///                 .await
///                 .expect("expecting successful response");
///
///             response.next_response().await;
///         }
///         .with_subscriber(assert_snapshot_subscriber!())
///         .await
///     }
/// ```
///
/// You can pass in a configuration and a schema to the test harness. If you pass in a schema, the test harness will create a query planner and use the schema to extract subgraph schemas.
///
///
pub(crate) struct PluginTestHarness<T: Plugin> {
    plugin: Box<dyn DynPlugin>,
    phantom: std::marker::PhantomData<T>,
}
#[buildstructor::buildstructor]
impl<T: Plugin> PluginTestHarness<T> {
    #[builder]
    pub(crate) async fn new<'a, 'b>(config: Option<&'a str>, schema: Option<&'b str>) -> Self {
        let factory = crate::plugin::plugins()
            .find(|factory| factory.type_id == TypeId::of::<T>())
            .expect("plugin not registered");

        let config = Configuration::from_str(config.unwrap_or_default())
            .expect("valid config required for test");

        let name = &factory.name.replace("apollo.", "");
        let config_for_plugin = config
            .validated_yaml
            .clone()
            .expect("invalid yaml")
            .as_object()
            .expect("invalid yaml")
            .get(name)
            .cloned()
            .unwrap_or(Value::Object(Default::default()));

        let (supergraph_sdl, parsed_schema, subgraph_schemas) = if let Some(schema) = schema {
            let planner = BridgeQueryPlanner::new(schema.to_string(), Arc::new(config))
                .await
                .unwrap();
            (
                schema.to_string(),
                planner.schema().supergraph_schema().clone(),
                planner.subgraph_schemas(),
            )
        } else {
            (
                "".to_string(),
                Valid::assume_valid(apollo_compiler::Schema::new()),
                Default::default(),
            )
        };

        let plugin_init = PluginInit::builder()
            .config(config_for_plugin.clone())
            .supergraph_sdl(Arc::new(supergraph_sdl))
            .supergraph_schema(Arc::new(parsed_schema))
            .subgraph_schemas(subgraph_schemas)
            .notify(Notify::default())
            .build();

        let plugin = factory
            .create_instance(plugin_init)
            .await
            .expect("failed to create plugin");

        Self {
            plugin,
            phantom: Default::default(),
        }
    }

    #[allow(dead_code)]
    pub(crate) async fn call_router(
        &self,
        request: router::Request,
        response_fn: fn(router::Request) -> router::Response,
    ) -> Result<router::Response, BoxError> {
        let service: router::BoxService = router::BoxService::new(
            ServiceBuilder::new()
                .service_fn(move |req: router::Request| async move { Ok((response_fn)(req)) }),
        );

        self.plugin.router_service(service).call(request).await
    }

    pub(crate) async fn call_supergraph(
        &self,
        request: supergraph::Request,
        response_fn: fn(supergraph::Request) -> supergraph::Response,
    ) -> Result<supergraph::Response, BoxError> {
        let service: supergraph::BoxService = supergraph::BoxService::new(
            ServiceBuilder::new()
                .service_fn(move |req: supergraph::Request| async move { Ok((response_fn)(req)) }),
        );

        self.plugin.supergraph_service(service).call(request).await
    }

    #[allow(dead_code)]
    pub(crate) async fn call_execution(
        &self,
        request: execution::Request,
        response_fn: fn(execution::Request) -> execution::Response,
    ) -> Result<execution::Response, BoxError> {
        let service: execution::BoxService = execution::BoxService::new(
            ServiceBuilder::new()
                .service_fn(move |req: execution::Request| async move { Ok((response_fn)(req)) }),
        );

        self.plugin.execution_service(service).call(request).await
    }

    #[allow(dead_code)]
    pub(crate) async fn call_subgraph(
        &self,
        request: subgraph::Request,
        response_fn: fn(subgraph::Request) -> subgraph::Response,
    ) -> Result<subgraph::Response, BoxError> {
        let name = request.subgraph_name.clone();
        let service: subgraph::BoxService = subgraph::BoxService::new(
            ServiceBuilder::new()
                .service_fn(move |req: subgraph::Request| async move { Ok((response_fn)(req)) }),
        );

        self.plugin
            .subgraph_service(&name.expect("subgraph name must be populated"), service)
            .call(request)
            .await
    }
    #[allow(dead_code)]
    pub(crate) async fn call_http_client(
        &self,
        subgraph_name: &str,
        request: http::HttpRequest,
        response_fn: fn(http::HttpRequest) -> http::HttpResponse,
    ) -> Result<http::HttpResponse, BoxError> {
        let service: http::BoxService = http::BoxService::new(
            ServiceBuilder::new()
                .service_fn(move |req: http::HttpRequest| async move { Ok((response_fn)(req)) }),
        );

        self.plugin
            .http_client_service(subgraph_name, service)
            .call(request)
            .await
    }
}

impl<T> Deref for PluginTestHarness<T>
where
    T: Plugin,
{
    type Target = T;

    fn deref(&self) -> &Self::Target {
        self.plugin
            .as_any()
            .downcast_ref()
            .expect("plugin should be of type T")
    }
}
