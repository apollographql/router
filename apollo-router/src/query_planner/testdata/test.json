{
  "data": {
    "hero": {
      "name": "R2-D2",
      "heroFriends": [
        {
          "id": "1000",
          "name": "Luke Skywalker"
        },
        {
          "id": "1002",
          "name": null
        },
        {
          "id": "1003",
          "name": "Leia Organa"
        }
      ]
    }
  },
  "errors": [
    {
      "message": "Name for character with ID 1002 could not be fetched.",
      "locations": [
        {
          "line": 6,
          "column": 7
        }
      ],
      "path": [
        {
          "String": "hero"
        },
        {
          "String": "heroFriends"
        },
        {
          "Number": 1
        },
        {
          "String": "name"
        }
      ],
      "extensions": {
        "error-extension": 5
      }
    }
  ],
  "extensions": {
    "response-extension": 3
  }
}
