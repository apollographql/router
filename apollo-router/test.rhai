/// The request's method
    // pub method: Method,

    // /// The request's URI
    // pub uri: Uri,

    // /// The request's version
    // pub version: Version,

    // /// The request's headers
    // pub headers: HeaderMap<HeaderValue>,

fn router_service() {
    request_val = "MYTESTINRHAISCRIPT";
    headers.coucou = "hello";
}