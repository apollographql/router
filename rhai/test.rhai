// At the supergraph_service stage responses.
fn router_service(service) {
    const response_callback = Fn("process_response"); // This is standard Rhai functionality for creating a function pointer
    service.map_response(response_callback); // Register the response callback
}

fn process_response(response) {
    if response.is_primary() {
        let len = response.body.errors.len();
        if response.body.errors > 0 {
            log_warn("error");

        }
        log_warn(len);
    }
}