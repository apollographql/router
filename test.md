this is a simple test
