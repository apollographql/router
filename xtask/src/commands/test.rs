use anyhow::Result;
use structopt::StructOpt;
use xtask::*;

const TEST_DEFAULT_ARGS: &[&str] = &["test", "--all", "--locked"];

#[derive(Debug, StructOpt)]
pub struct Test {}

impl Test {
    pub fn run(&self) -> Result<()> {
        eprintln!("Running tests");
        cargo!(TEST_DEFAULT_ARGS);

        #[cfg(windows)]
        {
            // dirty hack. Node processes on windows will not shut down cleanly.
            let _ = std::process::Command::new("taskkill")
                .args(["/f", "/im", "node.exe"])
                .spawn();
        }

        Ok(())
    }
}
