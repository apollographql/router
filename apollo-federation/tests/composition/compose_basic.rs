use insta::assert_snapshot;
use test_log::test;

use super::ServiceDefinition;
use super::compose_as_fed2_subgraphs;
use super::extract_subgraphs_from_supergraph_result;
use super::print_sdl; // TODO: Remove after rebasing onto merge commit of pull/8380

#[test]
fn generates_a_valid_supergraph() {
    let subgraph1 = ServiceDefinition {
        name: "Subgraph1",
        type_defs: r#"
        type Query {
          t: T
        }

        type T @key(fields: "k") {
          k: ID
        }

        type S {
          x: Int
        }

        union U = S | T
        "#,
    };

    let subgraph2 = ServiceDefinition {
        name: "Subgraph2",
        type_defs: r#"
        type T @key(fields: "k") {
          k: ID
          a: Int
          b: String
        }

        enum E {
          V1
          V2
        }
        "#,
    };

    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
    let supergraph = result.expect("Expected composition to succeed");
    let api_schema = supergraph
        .to_api_schema(Default::default())
        .expect("Expected API schema generation to succeed");

    assert_snapshot!(supergraph.schema().schema());
    assert_snapshot!(api_schema.schema());
}

#[test]
fn preserves_descriptions() {
    let subgraph1 = ServiceDefinition {
        name: "Subgraph1",
        type_defs: r#"
        "The foo directive description"
        directive @foo(url: String) on FIELD

        "A cool schema"
        schema {
          query: Query
        }

        """
        Available queries
        Not much yet
        """
        type Query {
          "Returns tea"
          t(
            "An argument that is very important"
            x: String!
          ): String
        }
        "#,
    };

    let subgraph2 = ServiceDefinition {
        name: "Subgraph2",
        type_defs: r#"
        "The foo directive description"
        directive @foo(url: String) on FIELD

        "An enum"
        enum E {
          "The A value"
          A
          "The B value"
          B
        }
        "#,
    };

    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
    let supergraph = result.expect("Expected composition to succeed");
    let api_schema = supergraph
        .to_api_schema(Default::default())
        .expect("Expected API schema generation to succeed");
    assert_snapshot!(api_schema.schema());
}

#[test]
fn no_hint_raised_when_merging_empty_description() {
    let subgraph1 = ServiceDefinition {
        name: "Subgraph1",
        type_defs: r#"
        schema {
          query: Query
        }

        ""
        type T {
          a: String @shareable
        }

        type Query {
          "Returns tea"
          t(
            "An argument that is very important"
            x: String!
          ): T
        }
        "#,
    };

    let subgraph2 = ServiceDefinition {
        name: "Subgraph2",
        type_defs: r#"
        "Type T"
        type T {
          a: String @shareable
        }
        "#,
    };

    let result = compose_as_fed2_subgraphs(&[subgraph1, subgraph2]);
    let supergraph = result.expect("Expected composition to succeed");

    // Verify that no hints are raised when merging empty description with non-empty description
    assert_eq!(
        supergraph.hints().len(),
        0,
        "Expected no hints but got: {:?}",
        supergraph.hints()
    );
}

#[test]
fn include_types_from_different_subgraphs() {
    let subgraph_a = ServiceDefinition {
        name: "subgraphA",
        type_defs: r#"
        type Query {
          products: [Product!]
        }

        type Product {
          sku: String!
          name: String!
        }
        "#,
    };

    let subgraph_b = ServiceDefinition {
        name: "subgraphB",
        type_defs: r#"
        type User {
          name: String
          email: String!
        }
        "#,
    };

    let result = compose_as_fed2_subgraphs(&[subgraph_a, subgraph_b]);
    let supergraph = result.expect("Expected composition to succeed");
    let api_schema = supergraph
        .to_api_schema(Default::default())
        .expect("Expected API schema generation to succeed");
    assert_snapshot!(api_schema.schema());

    // Validate extracted subgraphs contain proper federation directives
    let extracted_subgraphs = extract_subgraphs_from_supergraph_result(&supergraph)
        .expect("Expected subgraph extraction to succeed");

    let subgraph_a_extracted = extracted_subgraphs
        .get("subgraphA")
        .expect("Expected subgraphA to be present in extracted subgraphs");
    assert_snapshot!(subgraph_a_extracted.schema.schema());

    let subgraph_b_extracted = extracted_subgraphs
        .get("subgraphB")
        .expect("Expected subgraphB to be present in extracted subgraphs");
    assert_snapshot!(subgraph_b_extracted.schema.schema());
}

#[test]
fn doesnt_leave_federation_directives_in_the_final_schema() {
    let subgraph_a = ServiceDefinition {
        name: "subgraphA",
        type_defs: r#"
        type Query {
          products: [Product!] @provides(fields: "name")
        }

        type Product @key(fields: "sku") {
          sku: String!
          name: String! @external
        }
        "#,
    };

    let subgraph_b = ServiceDefinition {
        name: "subgraphB",
        type_defs: r#"
        type Product @key(fields: "sku") {
          sku: String!
          name: String! @shareable
        }
        "#,
    };

    let result = compose_as_fed2_subgraphs(&[subgraph_a, subgraph_b]);
    let supergraph = result.expect("Expected composition to succeed");
    let api_schema = supergraph
        .to_api_schema(Default::default())
        .expect("Expected API schema generation to succeed");
    assert_snapshot!(api_schema.schema());

    // Validate that federation directives (@provides, @key, @external, @shareable)
    // are properly rebuilt in the extracted subgraphs
    let extracted_subgraphs = extract_subgraphs_from_supergraph_result(&supergraph)
        .expect("Expected subgraph extraction to succeed");

    let subgraph_a_extracted = extracted_subgraphs
        .get("subgraphA")
        .expect("Expected subgraphA to be present in extracted subgraphs");
    assert_snapshot!(subgraph_a_extracted.schema.schema());

    let subgraph_b_extracted = extracted_subgraphs
        .get("subgraphB")
        .expect("Expected subgraphB to be present in extracted subgraphs");
    assert_snapshot!(subgraph_b_extracted.schema.schema());
}

#[test]
fn merges_default_arguments_when_they_are_arrays() {
    let subgraph_a = ServiceDefinition {
        name: "subgraph-a",
        type_defs: r#"
        type Query {
          a: A @shareable
        }

        type A @key(fields: "id") {
          id: ID
          get(ids: [ID] = []): [B] @external
          req: Int @requires(fields: "get { __typename }")
        }

        type B @key(fields: "id", resolvable: false) {
          id: ID
        }
        "#,
    };

    let subgraph_b = ServiceDefinition {
        name: "subgraph-b",
        type_defs: r#"
        type Query {
          a: A @shareable
        }

        type A @key(fields: "id") {
          id: ID
          get(ids: [ID] = []): [B]
        }

        type B @key(fields: "id") {
          id: ID
        }
        "#,
    };

    let result = compose_as_fed2_subgraphs(&[subgraph_a, subgraph_b]);
    let _supergraph = result.expect("Expected composition to succeed");
}
