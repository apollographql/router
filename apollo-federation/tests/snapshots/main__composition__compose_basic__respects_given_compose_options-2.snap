---
source: apollo-federation/tests/composition/compose_basic.rs
expression: print_sdl(api_schema.schema())
---
enum E {
  V1
  V2
}

type Query {
  t: T
}

type S {
  x: Int
}

type T {
  a: Int
  b: String
  k: ID
}

union U = S | T